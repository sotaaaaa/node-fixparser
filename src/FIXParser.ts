import { Socket } from 'node:net';
import {
  type ConnectionOptions,
  type KeyObject,
  connect as TLSConnect,
  type TLSSocket,
} from 'node:tls';
import { WebSocket } from 'ws';

import {
  type ConnectionType,
  FIXParserBase,
  type Options,
  type Protocol,
} from './FIXParserBase';
import type { IFIXParser } from './IFIXParser';
import { Field } from './fields/Field';
import * as Constants from './fieldtypes';
import { Field as FieldType } from './fieldtypes/Field';
import { Message as MessageType } from './fieldtypes/Message';
import { Message } from './message/Message';
import { heartBeat } from './messagetemplates/MessageTemplates';
import { clientProcessMessage } from './session/ClientMessageProcessor';
import { FrameDecoder } from './util/FrameDecoder';
import { Logger } from './util/Logger';
import { MessageBuffer } from './util/MessageBuffer';
import {
  DEFAULT_FIX_VERSION,
  DEFAULT_HEARTBEAT_SECONDS,
  type Parser,
  READY_MS,
  type Version,
  timestamp,
  version,
} from './util/util';

/**
 * FIXParser class.
 */
class FIXParser implements IFIXParser {
  /* FIXParser Version */
  public static version: Version = version;

  /* Name of current parser name */
  public parserName: Parser = 'FIXParser';
  /* Instance of FIXParser base class */
  public fixParserBase: FIXParserBase = new FIXParserBase();
  /* Next expected sequence number in */
  public nextNumIn = 1;
  /* Next expected sequence number out */
  public nextNumOut = 1;
  /* Heartbeat interval ID */
  public heartBeatIntervalId: ReturnType<typeof setInterval> | null = null;
  /* Type of socket connection (Socket, WebSocket, TLSSocket) */
  public socket: Socket | WebSocket | TLSSocket | null = null;
  /* Current connection status */
  public connected = false;
  /* Target hostname or IP address */
  public host: string | null = null;
  /* Target port */
  public port: number | null = null;
  /* Type of protocol */
  public protocol: Protocol | null = 'tcp';
  /* Assigned value used to identify firm sending message, SenderCompID tag 49 */
  public sender: string | null = null;
  /* Assigned value used to identify receiving firm, TargetCompID tag 56 */
  public target: string | null = null;
  /* Heartbeat interval in seconds */
  public heartBeatInterval: number = DEFAULT_HEARTBEAT_SECONDS;
  /* Protocol version, defaults to "FIXT.1.1" used by BeginString tag 8 */
  public fixVersion: string = DEFAULT_FIX_VERSION;
  /* Message buffer of 2500 incoming messages */
  public messageBufferIn: MessageBuffer = new MessageBuffer();
  /* Message buffer of 2500 outgoing messages */
  public messageBufferOut: MessageBuffer = new MessageBuffer();
  /* Whether an outgoing message contains a Logout */
  public requestedLogout = false;
  /* Indicates whether this instance is acceptor or initiator */
  public connectionType: ConnectionType = 'initiator';
  /* Logger instance */
  public logger: Logger = new Logger();

  /**
   * onMessageCallback is called when a message has been received
   *
   * @callback FIXParser~onMessageCallback
   * @param {Message} message
   */
  private onMessageCallback: Options['onMessage'] = message => {};

  /**
   * onOpenCallback is called the FIX connection has been initiated
   *
   * @callback FIXParser~onOpenCallback
   */
  private onOpenCallback: Options['onOpen'] = () => {};

  /**
   * onErrorCallback is called the FIX connection failed
   *
   * @callback FIXParser~onErrorCallback
   * @param {Error} error
   */
  private onErrorCallback: Options['onError'] = error => {};

  /**
   * onCloseCallback is called the FIX connection has been closed
   *
   * @callback FIXParser~onCloseCallback
   */
  private onCloseCallback: Options['onClose'] = () => {};

  /**
   * onReadyCallback is called the FIX connection is opened and ready
   *
   * @callback FIXParser~onReadyCallback
   */
  private onReadyCallback: Options['onReady'] = () => {};

  /**
   * Connect to a remote FIX server/gateway.
   * @param {Options} options - Connection options.
   */
  public connect(
    {
      host = 'localhost',
      port = 9878,
      protocol = 'tcp',
      sender = 'SENDER',
      target = 'TARGET',
      heartbeatIntervalSeconds = DEFAULT_HEARTBEAT_SECONDS,
      fixVersion = this.fixVersion,
      tlsKey = null,
      tlsCert = null,
      tlsUseSNI = false,
      tlsSkipStdInPipe = false,
      logging = true,
      logOptions = undefined,
      proxy = null,
      onMessage,
      onOpen,
      onError,
      onClose,
      onReady,
    }: Options = {
      onMessage: this.onMessageCallback,
      onOpen: this.onOpenCallback,
      onError: this.onErrorCallback,
      onClose: this.onCloseCallback,
      onReady: this.onReadyCallback,
    }
  ): void {
    this.logger = new Logger(
      logOptions
        ? {
            name: logOptions.name || 'fixparser',
            ...logOptions,
          }
        : {
            name: 'fixparser',
            format: 'json',
          }
    );
    if (!logging) {
      this.logger.silent = true;
    }
    this.connectionType = 'initiator';
    this.fixVersion = fixVersion;
    this.fixParserBase.fixVersion = fixVersion;
    this.protocol = protocol;
    this.sender = sender;
    this.target = target;
    this.heartBeatInterval = heartbeatIntervalSeconds;

    if (onMessage !== undefined) {
      this.onMessageCallback = onMessage;
    }

    if (onOpen !== undefined) {
      this.onOpenCallback = onOpen;
    }

    if (onError !== undefined) {
      this.onErrorCallback = onError;
    }

    if (onClose !== undefined) {
      this.onCloseCallback = onClose;
    }

    if (onReady !== undefined) {
      this.onReadyCallback = onReady;
    }

    if (protocol === 'tcp') {
      this.socket = new Socket();
      this.socket.setEncoding('ascii');
      this.socket.pipe(new FrameDecoder()).on('data', (data: string) => {
        const messages: Message[] = this.parse(data.toString());
        let i = 0;
        for (i; i < messages.length; i++) {
          clientProcessMessage(this, messages[i]);
          this.messageBufferIn.add(messages[i]);
          this.onMessageCallback?.(messages[i]);
        }
      });
      this.socket.connect(port, host, () => {
        this.connected = true;
        this.logger.log({
          level: 'info',
          message: `FIXParser (${this.protocol?.toUpperCase()}): -- Connected`,
        });
        this.onOpenCallback?.();
      });
      this.socket.on('close', () => {
        this.connected = false;
        this.onCloseCallback?.();
        this.stopHeartbeat();
      });
      this.socket.on('ready', () => {
        setTimeout(() => this.onReadyCallback?.(), READY_MS);
      });
      this.socket.on('timeout', () => {
        this.connected = false;
        const socket: Socket = this.socket! as Socket;
        this.onCloseCallback?.();
        socket.end();
        this.stopHeartbeat();
      });
      this.socket.on('error', error => {
        this.connected = false;
        this.onErrorCallback?.(error);
        this.stopHeartbeat();
      });
    } else if (protocol === 'websocket') {
      const connectionString: string =
        host.indexOf('ws://') === -1 && host.indexOf('wss://') === -1
          ? `ws://${host}:${port}`
          : `${host}:${port}`;
      if (proxy) {
        this.socket = new WebSocket(connectionString, { agent: proxy });
      } else {
        this.socket = new WebSocket(connectionString);
      }
      this.socket.on('message', (data: string | Buffer) => {
        const messages: Message[] = this.parse(data.toString());
        let i = 0;
        for (i; i < messages.length; i++) {
          clientProcessMessage(this, messages[i]);
          this.messageBufferIn.add(messages[i]);
          this.onMessageCallback?.(messages[i]);
        }
      });
      this.socket.on('open', () => {
        this.logger.log({
          level: 'info',
          message: `FIXParser (${this.protocol?.toUpperCase()}): -- Connected`,
        });
        this.connected = true;
        this.onOpenCallback?.();
      });
      this.socket.on('close', () => {
        this.connected = false;
        this.onCloseCallback?.();
        this.stopHeartbeat();
      });
      if (
        this.socket.readyState ===
        (WebSocket.OPEN || WebSocket.CLOSED || WebSocket.CLOSING || WebSocket.CONNECTING)
      ) {
        setTimeout(() => this.onReadyCallback?.(), READY_MS);
      }
    } else if (protocol === 'ssl-tcp' || protocol === 'tls-tcp') {
      const options: ConnectionOptions = {
        host,
        port,
        rejectUnauthorized: false,
      };

      if (tlsKey && tlsCert) {
        options.key = tlsKey as
          | string
          | Buffer
          | Array<string | Buffer | KeyObject>
          | undefined;
        options.cert = tlsCert as string | Buffer | Array<string | Buffer> | undefined;
      }

      if (tlsUseSNI) {
        options.servername = host;
      }

      this.socket = TLSConnect(port, host, options, () => {
        this.connected = true;
        this.onOpenCallback?.();
        this.logger.log({
          level: 'info',
          message: `FIXParser (${this.protocol?.toUpperCase()}): -- Connected through TLS`,
        });

        if (!tlsSkipStdInPipe) {
          process.stdin.pipe(this.socket as TLSSocket);
          process.stdin.resume();
        }
      });
      this.socket.setEncoding('utf8');
      (this.socket as TLSSocket).pipe(new FrameDecoder()).on('data', (data: string) => {
        const messages: Message[] = this.parse(data.toString());
        let i = 0;
        for (i; i < messages.length; i++) {
          clientProcessMessage(this, messages[i]);
          this.messageBufferIn.add(messages[i]);
          this.onMessageCallback?.(messages[i]);
        }
      });
      this.socket.on('close', () => {
        this.connected = false;
        this.onCloseCallback?.();
        this.stopHeartbeat();
      });
      this.socket.on('timeout', () => {
        const socket: TLSSocket = this.socket! as TLSSocket;
        this.connected = false;
        this.onCloseCallback?.();
        socket.end();
        this.stopHeartbeat();
      });
      this.socket.on('error', (error: Error) => {
        this.connected = false;
        this.onErrorCallback?.(error);
        this.stopHeartbeat();
      });
      setTimeout(() => this.onReadyCallback?.(), READY_MS);
    }
  }

  /**
   * Get the next outgoing message sequence number.
   *
   * @return {number} The next outgoing message sequence number.
   */
  public getNextTargetMsgSeqNum(): number {
    return this.nextNumOut;
  }

  /**
   * Set the next outgoing message sequence number.
   *
   * @param {number} nextMsgSeqNum - The next message sequence number.
   * @return {number} The next outgoing message sequence number.
   */
  public setNextTargetMsgSeqNum(nextMsgSeqNum: number): number {
    this.nextNumOut = nextMsgSeqNum;
    return this.nextNumOut;
  }

  /**
   * Get current timestamp.
   *
   * @param {Date} dateObject - An instance of a Date class.
   * @return {string} The current timestamp.
   */
  public getTimestamp(dateObject = new Date()): string {
    return timestamp(dateObject, this.logger);
  }

  /**
   * Create an instance of a FIX Message.
   *
   * @param {Array.<Field>} fields - An array of Fields.
   * @return {Message} A FIX Message class instance.
   */
  public createMessage(...fields: Field[]): Message {
    const message: Message = new Message(this.fixVersion, ...fields);
    message.messageSequence = this.getNextTargetMsgSeqNum();
    return message;
  }

  /**
   * Parse a FIX message string into Message instance(s).
   *
   * @param {string} data - FIX message string.
   * @return {Message} FIX Message class instance(s).
   */
  public parse(data: string): Message[] {
    return this.fixParserBase.parse(data);
  }

  /**
   * Send a FIX message.
   *
   * @param {Message} message - FIX Message class instance.
   */
  public send(message: Message): void {
    const messageType: Field | undefined = message.getField(FieldType.MsgType);
    if (messageType && messageType.value === MessageType.Logout) {
      this.requestedLogout = true;
    }

    const encodedMessage: string = message.encode();
    if (this.protocol === 'tcp' && this.connected) {
      this.setNextTargetMsgSeqNum(this.getNextTargetMsgSeqNum() + 1);
      (this.socket! as Socket).write(encodedMessage);
      this.messageBufferOut.add(message.clone());
      this.restartHeartbeat();
      this.logger.log({
        level: 'info',
        message: `FIXParser (${this.protocol.toUpperCase()}): >> sent`,
        fix: encodedMessage.replace(/\x01/g, '|'),
      });
    } else if (
      this.protocol === 'websocket' &&
      (this.socket! as WebSocket).readyState === WebSocket.OPEN
    ) {
      this.setNextTargetMsgSeqNum(this.getNextTargetMsgSeqNum() + 1);
      (this.socket! as WebSocket).send(encodedMessage);
      this.messageBufferOut.add(message.clone());
      this.restartHeartbeat();
      this.logger.log({
        level: 'info',
        message: `FIXParser (${this.protocol.toUpperCase()}): >> sent`,
        fix: encodedMessage.replace(/\x01/g, '|'),
      });
    } else if (
      (this.protocol === 'ssl-tcp' || this.protocol === 'tls-tcp') &&
      this.connected
    ) {
      this.setNextTargetMsgSeqNum(this.getNextTargetMsgSeqNum() + 1);
      (this.socket! as TLSSocket).write(encodedMessage);
      this.messageBufferOut.add(message.clone());
      this.restartHeartbeat();
      this.logger.log({
        level: 'info',
        message: `FIXParser (${this.protocol.toUpperCase()}): >> sent`,
        fix: encodedMessage.replace(/\x01/g, '|'),
      });
    } else {
      this.logger.logError({
        message: `FIXParser (${this.protocol?.toUpperCase()}): -- Could not send message, no connection`,
        fix: encodedMessage.replace(/\x01/g, '|'),
      });
    }
  }

  /**
   * Get connection status.
   *
   * @return {boolean} Current connection status.
   */
  public isConnected(): boolean {
    return this.connected;
  }

  /**
   * Close current connection.
   */
  public close(): void {
    if (this.protocol === 'tcp') {
      const socket: Socket = this.socket! as Socket;
      if (socket) {
        socket.destroy();
        this.connected = false;
      } else {
        this.logger.logError({
          message: `FIXParser (${this.protocol.toUpperCase()}): -- Could not close socket, connection not open`,
        });
      }
    } else if (this.protocol === 'websocket') {
      const socket: WebSocket = this.socket! as WebSocket;
      if (socket) {
        try {
          socket.close();
        } catch (error) {
          this.logger.logError({
            message: `FIXParser (${this.protocol.toUpperCase()}): -- Could not close socket,`,
            error,
          });
        }
        this.connected = false;
      } else {
        this.logger.logError({
          message: `FIXParser (${this.protocol.toUpperCase()}): -- Could not close socket, connection not open`,
        });
      }
    } else if (this.protocol === 'ssl-tcp' || this.protocol === 'tls-tcp') {
      const socket: TLSSocket = this.socket! as TLSSocket;
      if (socket) {
        socket.destroy();
        this.connected = false;
      } else {
        this.logger.logError({
          message: `FIXParser (${this.protocol.toUpperCase()}): -- Could not close socket, connection not open`,
        });
      }
    }
  }

  /**
   * Stop heartbeat interval.
   */
  public stopHeartbeat(): void {
    clearInterval(this.heartBeatIntervalId!);
  }

  /**
   * Restart heartbeat interval.
   */
  public restartHeartbeat(): void {
    this.stopHeartbeat();
    this.startHeartbeat(this.heartBeatInterval, true);
  }

  /**
   * Start heartbeat interval.
   *
   * @param {number} heartBeatInterval - Heartbeat interval in seconds.
   * @param {boolean} disableLog - Whether to disable heartbeat logs.
   */
  public startHeartbeat(
    heartBeatInterval: number = this.heartBeatInterval,
    disableLog?: boolean
  ): void {
    this.stopHeartbeat();
    if (!disableLog) {
      this.logger.log({
        level: 'info',
        message: `FIXParser (${this.protocol?.toUpperCase()}): -- Heartbeat configured to ${heartBeatInterval} seconds`,
      });
    }
    this.heartBeatInterval = heartBeatInterval;
    this.heartBeatIntervalId = setInterval(() => {
      const heartBeatMessage: Message = heartBeat(this);
      this.send(heartBeatMessage);
      const encodedMessage: string = heartBeatMessage.encode();
      this.logger.log({
        level: 'info',
        message: `FIXParser (${this.protocol?.toUpperCase()}): >> sent Heartbeat`,
        fix: encodedMessage.replace(/\x01/g, '|'),
      });
    }, this.heartBeatInterval * 1000);
  }
}

export * from './fieldtypes';
export type { Protocol } from './FIXParserBase';
export type { Options } from './FIXParserBase';
export type { MessageError } from './message/Message';
export { Constants };
export { Field };
export { Message };
export { FIXParser };
