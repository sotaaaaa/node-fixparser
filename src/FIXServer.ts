import { type Server, type Socket, createServer as createNetServer } from 'node:net';
import { type ServerOptions, type WebSocket, WebSocketServer } from 'ws';

import { FIXParser } from './FIXParser';
import type {
  ConnectionType,
  Options as FIXParserOptions,
  Protocol,
} from './FIXParserBase';
import type { IFIXParser } from './IFIXParser';
import { Field } from './fields/Field';
import * as Constants from './fieldtypes';
import { Message } from './message/Message';
import { heartBeat } from './messagetemplates/MessageTemplates';
import { serverProcessMessage } from './session/ServerMessageProcessor';
import { FrameDecoder } from './util/FrameDecoder';
import { Logger } from './util/Logger';
import { MessageBuffer } from './util/MessageBuffer';
import {
  DEFAULT_FIX_VERSION,
  DEFAULT_HEARTBEAT_SECONDS,
  type Parser,
  READY_MS,
  type Version,
  version,
} from './util/util';

type Options = Pick<
  FIXParserOptions,
  | 'host'
  | 'port'
  | 'protocol'
  | 'sender'
  | 'target'
  | 'heartbeatIntervalSeconds'
  | 'fixVersion'
  | 'logging'
  | 'logOptions'
  | 'onMessage'
  | 'onOpen'
  | 'onError'
  | 'onClose'
  | 'onReady'
>;

class FIXServer implements IFIXParser {
  /* FIXParser Version */
  public static version: Version = version;

  /* Name of current parser name */
  public parserName: Parser = 'FIXServer';
  /* Instance of FIXParser class */
  public fixParser: FIXParser = new FIXParser();
  /* Heartbeat interval ID */
  public heartBeatIntervalId: ReturnType<typeof setInterval> | null = null;
  /* Type of socket connection (Socket, WebSocket, TLSSocket) */
  public socket: WebSocket | Socket | null = null;
  /* Current connection status */
  public connected = false;
  /* Target hostname or IP address */
  public host = 'localhost';
  /* Target port */
  public port = 9878;
  /* Type of protocol */
  public protocol: Protocol = 'tcp';
  /* Assigned value used to identify firm sending message, SenderCompID tag 49 */
  public sender = '';
  /* Assigned value used to identify receiving firm, TargetCompID tag 56 */
  public target = '';
  /* Heartbeat interval in seconds */
  public heartBeatInterval: number = DEFAULT_HEARTBEAT_SECONDS;
  /* Protocol version, defaults to "FIXT.1.1" used by BeginString tag 8 */
  public fixVersion: string = DEFAULT_FIX_VERSION;
  /* Message buffer of 2500 incoming messages */
  public messageBufferIn: MessageBuffer = new MessageBuffer();
  /* Message buffer of 2500 outgoing messages */
  public messageBufferOut: MessageBuffer = new MessageBuffer();
  /* Whether an outgoing message contains a Logout */
  public requestedLogout = false;
  /* Indicates whether this instance is acceptor or initiator */
  public connectionType: ConnectionType = 'acceptor';
  /* Logger instance */
  public logger: Logger = new Logger();
  /* Server socket connection */
  public server: Server | WebSocketServer | null = null;
  /* Next expected sequence number in */
  public nextNumIn = 1;
  /* Whether current FIX session has logged in */
  public isLoggedIn = false;
  /* Message counter */
  public messageCounter = 0;

  // Required logon fields
  // Example: Add fields required for logon (username, password, etc.)
  public requiredLogonFields: number[] = [];

  /**
   * onMessageCallback is called when a message has been received
   *
   * @callback FIXParser~onMessageCallback
   * @param {Message} message
   */
  private onMessageCallback: Options['onMessage'] = message => {};

  /**
   * onOpenCallback is called the FIX connection has been initiated
   *
   * @callback FIXParser~onOpenCallback
   */
  private onOpenCallback: Options['onOpen'] = () => {};

  /**
   * onErrorCallback is called the FIX connection failed
   *
   * @callback FIXParser~onErrorCallback
   * @param {Error} error
   */
  private onErrorCallback: Options['onError'] = error => {};

  /**
   * onCloseCallback is called the FIX connection has been closed
   *
   * @callback FIXParser~onCloseCallback
   */
  private onCloseCallback: Options['onClose'] = () => {};

  /**
   * onReadyCallback is called the FIX connection is opened and ready
   *
   * @callback FIXParser~onReadyCallback
   */
  private onReadyCallback: Options['onReady'] = () => {};

  /**
   * Create FIX server/gateway.
   * @param {Options} options - Server options.
   */
  public createServer(
    {
      host = this.host,
      port = this.port,
      protocol = this.protocol,
      sender = this.sender,
      target = this.target,
      heartbeatIntervalSeconds = DEFAULT_HEARTBEAT_SECONDS,
      fixVersion = this.fixVersion,
      logging = true,
      logOptions = undefined,
      onMessage,
      onOpen,
      onError,
      onClose,
      onReady,
    }: Options = {
      onMessage: this.onMessageCallback,
      onOpen: this.onOpenCallback,
      onError: this.onErrorCallback,
      onClose: this.onCloseCallback,
      onReady: this.onReadyCallback,
    }
  ): void {
    this.logger = new Logger(
      logOptions
        ? {
            name: logOptions.name || 'fixserver',
            ...logOptions,
          }
        : {
            name: 'fixserver',
            format: 'json',
          }
    );
    if (!logging) {
      this.logger.silent = true;
    }
    this.connectionType = 'acceptor';
    this.host = host;
    this.port = port;
    this.protocol = protocol;
    this.sender = sender;
    this.target = target;
    this.fixParser.sender = sender;
    this.fixParser.target = target;
    this.heartBeatInterval = heartbeatIntervalSeconds;
    this.fixVersion = fixVersion;

    if (onMessage !== undefined) {
      this.onMessageCallback = onMessage;
    }

    if (onOpen !== undefined) {
      this.onOpenCallback = onOpen;
    }

    if (onError !== undefined) {
      this.onErrorCallback = onError;
    }

    if (onClose !== undefined) {
      this.onCloseCallback = onClose;
    }

    if (onReady !== undefined) {
      this.onReadyCallback = onReady;
    }

    this.initialize();
  }

  private initialize() {
    this.messageCounter = 0;
    if (this.protocol === 'tcp') {
      this.server = createNetServer((socket: Socket) => {
        this.socket = socket;
        this.socket.pipe(new FrameDecoder()).on('data', (data: string) => {
          this.connected = true;
          const messages: Message[] = this.parse(data.toString());
          let i = 0;
          for (i; i < messages.length; i++) {
            serverProcessMessage(this, messages[i]);
            this.messageBufferIn.add(messages[i]);
            this.onMessageCallback?.(messages[i]);
          }
        });
        this.socket.on('connect', () => {
          this.logger.log({
            level: 'info',
            message: `FIXServer (${this.protocol.toUpperCase()}): -- Connection established`,
          });
          this.connected = true;
          this.onOpenCallback?.();
        });
        this.socket.on('close', () => {
          this.connected = false;
          this.stopHeartbeat();
          this.resetSession();
          this.logger.log({
            level: 'info',
            message: `FIXServer (${this.protocol.toUpperCase()}): -- Closed connection`,
          });
          this.onCloseCallback?.();
        });
        this.socket.on('timeout', () => {
          this.connected = false;
          this.stopHeartbeat();
          this.close();
          this.resetSession();
          this.logger.logError({
            message: `FIXServer (${this.protocol.toUpperCase()}): -- Connection timeout`,
          });
          this.onCloseCallback?.();
        });
        this.socket.on('error', (error: Error) => {
          this.connected = false;
          this.stopHeartbeat();
          this.close();
          this.resetSession();
          this.logger.logError({
            message: `FIXServer (${this.protocol.toUpperCase()}): -- Error`,
            error,
          });
          this.onErrorCallback?.(error);
        });
      });
      this.server.listen(this.port, this.host, () => {
        this.logger.log({
          level: 'info',
          message: `FIXServer (${this.protocol.toUpperCase()}): -- Listening for connections at ${
            this.host
          }:${this.port}...`,
        });
        setTimeout(() => this.onReadyCallback?.(), READY_MS);
      });
    } else if (this.protocol === 'websocket') {
      const serverOptions: ServerOptions = {
        host: this.host,
        port: this.port,
      };
      this.server = new WebSocketServer(serverOptions);
      this.server.on('connection', socket => {
        this.connected = true;
        socket.on('message', (data: string | Buffer) => {
          const messages: Message[] = this.parse(data.toString());
          let i = 0;
          for (i; i < messages.length; i++) {
            serverProcessMessage(this, messages[i]);
            this.messageBufferIn.add(messages[i]);
            this.onMessageCallback?.(messages[i]);
          }
        });
      });
      this.server.on('close', () => {
        this.connected = false;
        this.stopHeartbeat();
        this.onCloseCallback?.();
        this.logger.log({
          level: 'info',
          message: `FIXServer (${this.protocol.toUpperCase()}): -- Closed connection`,
        });
      });
      this.server.on('error', error => {
        this.connected = false;
        this.stopHeartbeat();
        this.close();
        this.onErrorCallback?.(error);
        this.logger.logError({
          message: `FIXServer (${this.protocol.toUpperCase()}): -- Error`,
          error,
        });
      });
      this.server.on('listening', () => {
        this.logger.log({
          level: 'info',
          message: `FIXServer (${this.protocol.toUpperCase()}): -- Listening for connections at ${
            this.host
          }:${this.port}...`,
        });
        setTimeout(() => this.onReadyCallback?.(), READY_MS);
      });
    } else {
      this.logger.logError({
        message: `FIXServer: Create server, invalid protocol: ${this.protocol.toUpperCase()}`,
      });
    }
  }

  public getNextTargetMsgSeqNum(): number {
    return this.fixParser.getNextTargetMsgSeqNum();
  }

  public setNextTargetMsgSeqNum(nextMsgSeqNum: number): number {
    return this.fixParser.setNextTargetMsgSeqNum(nextMsgSeqNum);
  }

  public getTimestamp(dateObject = new Date()): string {
    return this.fixParser.getTimestamp(dateObject);
  }

  public createMessage(...fields: Field[]): Message {
    return this.fixParser.createMessage(...fields);
  }

  public parse(data: string): Message[] {
    return this.fixParser.parse(data);
  }

  public send(message: Message): void {
    if (this.protocol === 'tcp') {
      const socket: Socket = this.socket! as Socket;
      const encodedMessage: string = message.encode();
      this.fixParser.setNextTargetMsgSeqNum(this.fixParser.getNextTargetMsgSeqNum() + 1);
      if (!socket.write(encodedMessage)) {
        this.logger.logError({
          message: `FIXServer (${this.protocol.toUpperCase()}): -- Could not send message, socket not open`,
        });
      } else {
        this.messageBufferOut.add(message.clone());
        this.restartHeartbeat();
        this.logger.log({
          level: 'info',
          message: `FIXServer (${this.protocol.toUpperCase()}): >> sent`,
          fix: encodedMessage.replace(/\x01/g, '|'),
        });
      }
    } else if (this.protocol === 'websocket') {
      const server: WebSocketServer = this.server! as WebSocketServer;
      const encodedMessage: string = message.encode();
      if (server?.clients && server.clients.size > 0) {
        server.clients.forEach((client: WebSocket) => {
          if (client.readyState === client.OPEN) {
            this.fixParser.setNextTargetMsgSeqNum(
              this.fixParser.getNextTargetMsgSeqNum() + 1
            );
            client.send(encodedMessage);
            this.messageBufferOut.add(message.clone());
            this.restartHeartbeat();
            this.logger.log({
              level: 'info',
              message: `FIXServer (${this.protocol.toUpperCase()}): >> sent`,
              fix: encodedMessage.replace(/\x01/g, '|'),
            });
          }
        });
      } else {
        this.logger.logError({
          message: `FIXServer (${this.protocol.toUpperCase()}): -- Could not send message, socket not connected`,
          fixMessage: message,
        });
      }
    }
  }

  /**
   * Get connection status.
   *
   * @return {boolean} Current connection status.
   */
  public isConnected(): boolean {
    return this.connected;
  }

  /**
   * Resets the current session.
   */
  private resetSession() {
    this.isLoggedIn = false;
    this.messageCounter = 0;
  }

  /**
   * Close current connection.
   */
  public close(): void {
    if (this.protocol === 'tcp') {
      const socket: Socket = this.socket! as Socket;
      const server: Server = this.server! as Server;
      if (socket?.end) {
        socket.end(() => {
          if (server) {
            server.close(() => {
              this.logger.log({
                level: 'info',
                message: `FIXServer (${this.protocol.toUpperCase()}): -- Ended session`,
              });
              this.initialize();
            });
          }
        });
      }
    } else if (this.protocol === 'websocket') {
      const server: WebSocketServer = this.server! as WebSocketServer;
      if (server) {
        server.clients.forEach((client: WebSocket) => {
          client.close();
        });
        server.close(() => {
          this.logger.log({
            level: 'info',
            message: `FIXServer (${this.protocol.toUpperCase()}): -- Ended session`,
          });
          this.initialize();
        });
      }
    }
  }

  public destroy(): void {
    this.stopHeartbeat();
    if (this.protocol === 'tcp') {
      const socket: Socket = this.socket! as Socket;
      const server: Server = this.server! as Server;
      socket.destroy();
      server.close(() => {
        this.logger.log({
          level: 'info',
          message: `FIXServer (${this.protocol.toUpperCase()}): -- Destroyed`,
        });
      });
    } else if (this.protocol === 'websocket') {
      const server: WebSocketServer = this.server! as WebSocketServer;
      if (server) {
        server.clients.forEach((client: WebSocket) => {
          client.close();
        });
        server.close(() => {
          this.logger.log({
            level: 'info',
            message: `FIXServer (${this.protocol.toUpperCase()}): -- Destroyed`,
          });
        });
      }
    }
  }

  /**
   * Stop heartbeat interval.
   */
  public stopHeartbeat(): void {
    clearInterval(this.heartBeatIntervalId!);
  }

  /**
   * Restart heartbeat interval.
   */
  public restartHeartbeat(): void {
    this.stopHeartbeat();
    this.startHeartbeat();
  }

  /**
   * Start heartbeat interval.
   *
   * @param {number} heartBeatInterval - Heartbeat interval in seconds.
   * @param {boolean} disableLog - Whether to disable heartbeat logs.
   */
  public startHeartbeat(
    heartBeatInterval: number = this.heartBeatInterval,
    disableLog?: boolean
  ): void {
    this.stopHeartbeat();
    if (!disableLog) {
      this.logger.log({
        level: 'info',
        message: `FIXServer (${this.protocol.toUpperCase()}): -- Heartbeat configured to ${heartBeatInterval} seconds`,
      });
    }
    this.heartBeatInterval = heartBeatInterval;
    this.heartBeatIntervalId = setInterval(() => {
      const heartBeatMessage: Message = heartBeat(this);
      this.send(heartBeatMessage);
      const encodedMessage: string = heartBeatMessage.encode();
      this.logger.log({
        level: 'info',
        message: `FIXServer (${this.protocol.toUpperCase()}): >> sent Heartbeat`,
        fix: encodedMessage.replace(/\x01/g, '|'),
      });
    }, this.heartBeatInterval * 1000);
  }
}

export * from './fieldtypes';
export type { Protocol } from './FIXParserBase';
export type { Options } from './FIXParserBase';
export type { MessageError } from './message/Message';
export { Constants };
export { Field };
export { Message };
export { FIXServer };
