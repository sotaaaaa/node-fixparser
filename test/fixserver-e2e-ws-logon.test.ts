import { FIXParser } from '../src/FIXParser';
import { EncryptMethod, FIXServer, Field, Fields, type Message, Messages } from '../src/FIXServer';

jest.setTimeout(30000);

describe('FIXServer WebSocket', () => {
    afterEach(() => {
        jest.clearAllMocks();
        jest.restoreAllMocks();
    });
    it('End-to-end: connect and Logon', (done) => {
        const fixServer: FIXServer = new FIXServer();
        const fixParser: FIXParser = new FIXParser();
        const HOST: string = 'localhost';

        // Start up a server
        fixServer.createServer({
            host: HOST,
            port: 9905,
            protocol: 'websocket',
            sender: 'SERVER2',
            target: 'CLIENT2',
            logging: false,
            onMessage: (message: Message) => {
                expect(message.description).toEqual('Logon');
                expect(message.messageString).toMatchSnapshot();
                expect(fixServer.heartBeatInterval).toEqual(128);
                expect(fixServer.fixVersion).toEqual('FIX.4.6');
                fixParser.close();
                fixServer.destroy();
                done();
            },
            onReady: () => {
                // Connect with a client
                fixParser.connect({
                    host: HOST,
                    port: 9905,
                    protocol: 'websocket',
                    sender: 'CLIENT2',
                    target: 'SERVER2',
                    fixVersion: 'FIX.4.6',
                    logging: false,
                    onOpen: () => {
                        // Send a Logon message
                        const logon: Message = fixParser.createMessage(
                            new Field(Fields.MsgType, Messages.Logon),
                            new Field(Fields.MsgSeqNum, fixParser.getNextTargetMsgSeqNum()),
                            new Field(Fields.SenderCompID, fixParser.sender),
                            new Field(Fields.SendingTime, fixParser.getTimestamp()),
                            new Field(Fields.TargetCompID, fixParser.target),
                            new Field(Fields.ResetSeqNumFlag, 'Y'),
                            new Field(Fields.EncryptMethod, EncryptMethod.None),
                            new Field(Fields.HeartBtInt, 128),
                        );
                        fixParser.send(logon);
                    },
                    onError: (error?: Error) => console.log('FIXParser: ', error),
                });
            },
            onError: (error?: Error) => console.log('FIXServer: ', error),
        });
    });
});
