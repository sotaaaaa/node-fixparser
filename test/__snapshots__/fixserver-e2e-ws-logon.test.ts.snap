// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FIXServer WebSocket End-to-end: connect and Logon 1`] = `"8=FIX.4.69=7635=A34=149=CLIENT252=20210815-21:51:47.36556=SERVER2141=Y98=0108=12810=105"`;
